src/algebra/lattice_ordered_group.lean

```diff
@@ -0,0 +1,595 @@
+{ mul_le_mul := λ a b hab c d, mul_le_mul' (mul_le_mul' le_rfl hab) le_rfl }
+
+
+-- A linearly ordered additive commutative group is a lattice ordered commutative group
+@[priority 100, to_additive] -- see Note [lower instance priority]
+instance linear_ordered_comm_group.to_lattice_ordered_comm_group (α : Type u)
+ [s : linear_ordered_comm_group α] : lattice_ordered_comm_group α :=
+{ mul_le_mul_left := s.mul_le_mul_left, }
+
+variables {α : Type u} [lattice_ordered_add_comm_group α]
+
+-- Special case of Bourbaki A.VI.9 (1)
+/--
+Let `α` be a lattice ordered commutative group. For all elements `a`, `b` and `c` in `α`,
+$$c + (a⊔b) = (c+a)⊔(c+b).$$
+-/
+lemma add_sup_eq_add_sup_add (a b c : α) : c + (a⊔b) = (c+a)⊔(c+b) :=
+begin
+  refine le_antisymm _ (by simp),
+  rw [← add_le_add_iff_left (-c), ← add_assoc, neg_add_self, zero_add],
+  apply sup_le,
+  { simp },
+  { simp },
+end
+
+-- Special case of Bourbaki A.VI.9 (2)
+/--
+Let `α` be a lattice ordered commutative group. For all elements `a` and `b` in `α`,
+$$-(a⊔b)=(-a)⊓(-b).$$
+-/
+lemma neg_sup_eq_neg_inf_neg (a b : α) : -(a⊔b)=(-a)⊓(-b) :=
+begin
+  rw le_antisymm_iff,
+  split,
+  { rw le_inf_iff,
+    split,
+    { rw neg_le_neg_iff, apply le_sup_left, },
+    { rw neg_le_neg_iff, apply le_sup_right, } },
+  { rw ← neg_le_neg_iff, simp,
+    split,
+    { rw ← neg_le_neg_iff, simp, },
+    { rw ← neg_le_neg_iff, simp, } }
+end
+
+/--
+Let `α` be a lattice ordered commutative group. For all elements `a` and `b` in `α`,
+$$-a⊔-b = -(a⊓b).$$
+-/
+@[simp] lemma sup_neg_eq_neg_inf (a b : α) : -a⊔-b = -(a⊓b) :=
+begin
+  rw [← neg_neg (-a⊔-b), neg_sup_eq_neg_inf_neg (-a) (-b), neg_neg, neg_neg],
+end
+
+-- Bourbaki A.VI.10 Prop 7
+/--
+Let `α` be a lattice ordered commutative group. For all elements `a` and `b` in `α`,
+$$a⊓b + (a⊔b) = a + b.$$
+-/
+lemma add_eq_inf_add_sup (a b : α) : a⊓b + (a⊔b) = a + b :=
+calc a⊓b + (a⊔b) = a⊓b + ((a + b) + ((-b)⊔(-a))) :
+  by {  rw add_sup_eq_add_sup_add (-b) (-a) (a+b), simp,  }
+... = a⊓b + ((a + b) + -(a⊓b)) : by { rw [← sup_neg_eq_neg_inf, sup_comm], }
+... = a + b                    : by { simp, }
+
+-- Bourbaki A.VI.12 Definition 4
+/--
+Let `α` be a lattice ordered commutative group with identity `0`. For an element `a` of type `α`,
+the element `a ⊔ 0` is said to be the *positive component* of `a`, denoted `a⁺`.
+-/
+def lattice_ordered_add_comm_group.pos (a : α) : α :=  a ⊔ 0
+postfix `⁺`:1000 := lattice_ordered_add_comm_group.pos
+
+/--
+Let `α` be a lattice ordered commutative group with identity `0`. For an element `a` of type `α`,
+the element `(-a) ⊔ 0` is said to be the *negative component* of `a`, denoted `a⁻`.
+-/
+def lattice_ordered_add_comm_group.neg (a : α) : α :=  (-a) ⊔ 0
+postfix `⁻`:1000 := lattice_ordered_add_comm_group.neg
+
+/--
+Absolute value is a unary operator with properties similar to the absolute value of a real number.
+-/
+class has_abs (α : Type u) := (abs : α → α)
```

Comment:
Why define a completely separate class, as opposed to a function that notation (possibly localized)?

Reply:
I think this typeclass is probably worth having, as we have lots of different definitions of `abs`. Maybe this is just `has_norm.norm` though.

Reply:
Ultimately I want to study lattice ordered groups equipped with a norm satisfying the "solid" property e.g.
```
class normed_lattice_add_comm_group (α : Type u)
  extends normed_group α, lattice α :=
(add_le_add_left : ∀ a b : α, a ≤ b → ∀ c : α, c + a ≤ c + b)
(notation `|`a`|` := a⊔-a) -- temporary notation within this definition
(solid' : ∀ a b : α, |a| ≤ |b| → ∥a∥ ≤ ∥b∥)
```
so it's necessary to distinguish between the absolute value (in α) and the norm (in the real field).

See the discussion at https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/norm.20of.20abs

(Probably covariant_class should be used in the above?)

Reply:
> I think this typeclass is probably worth having, as we have lots of different definitions of `abs`. Maybe this is just `has_norm.norm` though.

`norm` is hard-coded to take values in the reals. So `has_abs` would do something different. I agree that it should probably be added. But it would be better to split off a PR that does that.

Reply:
> But it would be better to split off a PR that does that.

By that, do you mean a PR to add it to [core](https://github.com/leanprover-community/lean/blob/master/library/init/core.lean )?

Reply:
@mans0954 No, I just mean splitting off a new PR to mathlib. It's good practice to make multiple PRs that all do one thing at a time.

Reply:
So the thing that this new PR would be going would be retrofitting `has_abs` to existing classes that have a notion of an absolute value?

Reply:
Yes. The new PR would introduces `has_abs`, and refactor the existing `abs` definition to make use of that class. I expect that this will mean 1 new file, and ~ 100 lines of changes. Which is certainly enough for a new PR.
The upshot of splitting is: if one of the two pieces is ready, that PR can be merged. If you keep everything together, then the reviewers constantly have to remember "this part of the PR was ok, but that file still needs a bit of work".

Reply:
These seem to be the current locations where something called `abs` is defined in mathlib

https://github.com/leanprover-community/mathlib/blob/f36c98e877dd86af12606abbba5275513baa8a26/src/algebra/ordered_group.lean#L984

https://github.com/leanprover-community/mathlib/blob/f36c98e877dd86af12606abbba5275513baa8a26/src/data/complex/is_R_or_C.lean#L439

https://github.com/leanprover-community/mathlib/blob/f36c98e877dd86af12606abbba5275513baa8a26/src/data/complex/basic.lean#L406

https://github.com/leanprover-community/mathlib/blob/f36c98e877dd86af12606abbba5275513baa8a26/src/data/real/nnreal.lean#L762

https://github.com/leanprover-community/mathlib/blob/f36c98e877dd86af12606abbba5275513baa8a26/src/data/num/basic.lean#L315

https://github.com/leanprover-community/mathlib/blob/f36c98e877dd86af12606abbba5275513baa8a26/src/topology/continuous_function/basic.lean#L113

Reply:
@jcommelin  I'm struggling to picture what would go in the new file, other than something like:
```
class has_abs (α : Type*) := (abs : α → α)
notation `|`a`|` := has_abs.abs a
```
Or would you propose adding some axioms (positive definite, triangle inequality, maybe |a| as a sup for -a and a) to a partially ordered group and trying to prove something like ||b|-|a|| ≤ |b-a|?

Reply:
Hmm, from you examples, only the first and the last have type `X -> X`. I had expected a bit more examples.

I think the new file would be very short, indeed. Just the things you wrote. But you can then import it in `algebra/ordered_group.lean` and in `topology/continuous_function/basic.lean`, and adapt those two examples. It would be good to also use the notation for `abs` in those files.

Reply:
@jcommelin I've been looking at this https://github.com/leanprover-community/mathlib/tree/absolute-value - but it's a big job - perhaps unsurprisingly `abs` crops up a lot in maths. I'm happy to keep going if this is what's required, but though I ought to check that this is really what's wanted before I go much further.

Reply:
In order to unblock this PR, I'm wondering about uncoupling it from #8891 by changing the definition of `abs` in `algebra/ordered_group` to:
```
@[to_additive abs]
def mabs {α : Type*} [has_inv α] [lattice α] (a : α) : α := a ⊔ (a⁻¹)
```
This breaks the use of `simp [abs]` elsewhere in mathlib, so one also needs something like:
```
lemma abs_eq_max_neg {α : Type*} [has_neg α] [linear_order α] (a : α) : abs a = max a (-a) :=
begin
  exact rfl,
end
```
This might be a simpler solution in the short term, although I think we will still want a `has_abs` class in the longer term as absolute values can arise in other contexts (e.g. the self-adjoint part of a C*-algebra and its dual).

Reply:
Something subtle seems to go wrong with `const_abs` in `order/filter/filter_product`. I tried fixing it with
```
lemma abs_def [linear_ordered_add_comm_group β] (x : β*) : abs x = map abs x :=
induction_on x $ λ a, by exact rfl
...
@[simp] lemma const_abs [linear_ordered_add_comm_group β] (x : β) :
  (↑(abs x) : β*) = abs ↑x :=
by rw [abs_def, map_const]
```
which lean is okay with, but then `coe_abs` in `data/real/hypereal` is broken.

Reply:
Doing `convert germ.const_abs x` on `coe_abs` in `data/real/hypereal`, it looks like I need to show `lattice_of_linear_order = germ.lattice`.

Reply:
This was resolved as explained [here](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/lattices.2C.20linear.20orders.20and.20germs)

Reply:
I've now uncoupled this PR from #8891 by making the definition of the `abs` function in `algebra/ordered_group`.